libs/s25main/desktops/dskGameInterface.cpp

```diff
@@ -692,37 +523,243 @@ bool dskGameInterface::Msg_LeftDown(const MouseCoords& mc)
+                         COLOR_YELLOW);
+        iconPos -= DrawPoint(magnifierImg->getWidth() + 4, 0);
+    }
+}
+
+bool dskGameInterface::Msg_LeftDown(const MouseCoords& mc)
+{
+    DrawPoint btOrig(VIDEODRIVER.GetRenderSize().x / 2 - LOADER.GetImageN("resource", 29)->getWidth() / 2 + 44,
+                     VIDEODRIVER.GetRenderSize().y - LOADER.GetImageN("resource", 29)->getHeight() + 4);
+    Extent btSize = Extent(37, 32) * 4u;
+    if(IsPointInRect(mc.GetPos(), Rect(btOrig, btSize)))
+        return false;
+
+    if(!VIDEODRIVER.IsTouch())
+    {
+        // Start scrolling also on Ctrl + left click
+        if(VIDEODRIVER.GetModKeyState().ctrl)
+        {
+            Msg_RightDown(mc);
+            return true;
+        } else if(isScrolling)
+            StopScrolling();
+
+        return ContextClick(mc);
+
+    } else if(mc.num_tfingers < 2)
+        touchDuration = VIDEODRIVER.GetTickCount();
+    else if(isScrolling) // Bei 2 Fingern soll gezoomt werden kein klick oder bewegen der Karte
+        StopScrolling();
+
+    return true;
+}
+
+bool dskGameInterface::Msg_LeftUp(const MouseCoords& mc)
 {
     if(isScrolling)
     {
         StopScrolling();
         return true;
     }
+    // num_tfingers wird erst reduziert nachdem diese Funktion läuft :/
+    // Herausfinden ob kurz genug unten um context click auszuführen
+    if(mc.num_tfingers == 1 && (VIDEODRIVER.GetTickCount() - touchDuration) < 250)
+        return ContextClick(mc);
+
     return false;
 }
 
 bool dskGameInterface::Msg_MouseMove(const MouseCoords& mc)
 {
     if(!isScrolling)
-        return false;
+    {
+        if(mc.num_tfingers == 1)
+            Msg_RightDown(mc);
+        else
+            return false;
+    }
+
+    // Natural scrolling
+    if(SETTINGS.interface.mouseMode == 2)
+    {
+        const Position mapPos = gwv.ViewPosToMap(mc.GetPos());
+        gwv.MoveBy(-(mapPos - startScrollPt));
+        startScrollPt = mapPos;
+    } else
+    {
+        int acceleration = SETTINGS.global.smartCursor ? 2 : 3;
 
-    int acceleration = SETTINGS.global.smartCursor ? 2 : 3;
+        if(SETTINGS.interface.mouseMode == 1)
+            acceleration = -acceleration;
 
-    if(SETTINGS.interface.invertMouse)
-        acceleration = -acceleration;
+        gwv.MoveBy((mc.GetPos() - startScrollPt) * acceleration);
+        VIDEODRIVER.SetMousePos(startScrollPt);
 
-    gwv.MoveBy((mc.GetPos() - startScrollPt) * acceleration);
-    VIDEODRIVER.SetMousePos(startScrollPt);
+        if(!SETTINGS.global.smartCursor)
+            startScrollPt = mc.GetPos();
+    }
 
-    if(!SETTINGS.global.smartCursor)
-        startScrollPt = mc.GetPos();
     return true;
 }
 
 bool dskGameInterface::Msg_RightDown(const MouseCoords& mc)
 {
-    StartScrolling(mc.pos);
+    if(SETTINGS.interface.mouseMode == 2)
+        StartScrolling(gwv.ViewPosToMap(mc.GetPos()));
+    else
+        StartScrolling(mc.pos);
```

Comment:
what I do not understand is the difference - is `mc.pos` and `mc.GetPos` not the same?

Reply:
Indeed. And `mc.pos` should be used. I created a PR to avoid the confusion: It is a struct and having a method for getting the position is a relict of the time where we didn't have mc.pos being a Point: https://github.com/Return-To-The-Roots/s25client/pull/1811